modules/gapi/src/backends/fluid/gfluidcore.cpp

```diff
@@ -151,6 +151,319 @@ GAPI_FLUID_KERNEL(GFluidAddW, cv::gapi::core::GAddW, false)
 
 enum Arithm { ARITHM_ABSDIFF, ARITHM_ADD, ARITHM_SUBTRACT, ARITHM_MULTIPLY, ARITHM_DIVIDE };
 
+#if CV_SIMD
+CV_ALWAYS_INLINE void absdiff_store(short out[], const v_int16& a, const v_int16& b, int x)
+{
+    vx_store(&out[x], v_absdiffs(a, b));
+}
+
+CV_ALWAYS_INLINE void absdiff_store(ushort out[], const v_uint16& a, const v_uint16& b, int x)
+{
+    vx_store(&out[x], v_absdiff(a, b));
+}
+
+CV_ALWAYS_INLINE void absdiff_store(uchar out[], const v_uint8& a, const v_uint8& b, int x)
+{
+    vx_store(&out[x], v_absdiff(a, b));
+}
+
+CV_ALWAYS_INLINE void absdiff_store(float out[], const v_float32& a, const v_float32& b, int x)
+{
+    vx_store(&out[x], v_absdiff(a, b));
+}
+
+template<typename T, typename VT>
+CV_ALWAYS_INLINE int absdiff_impl(const T in1[], const T in2[], T out[], int length)
+{
+    constexpr int nlanes = static_cast<int>(VT::nlanes);
+
+    if (length < nlanes)
+        return 0;
+
+    int x = 0;
+
+    for (; x < length; )
+    {
+        for (; x <= length - nlanes; x += nlanes)
+        {
+            VT a = vx_load(&in1[x]);
+            VT b = vx_load(&in2[x]);
+            absdiff_store(out, a, b, x);
+        }
+
+        if (x < length)
+            x = length - nlanes;
```

Comment:
Such tail processing requires that in != out. (However that's ok if such a check is performed somewhere higher by call stack)

Reply:
Ok. Add check for cases when input and output types are the same.

Reply:
@dmatveev Do we really need these checks in G-API code? Does G-API support inplace processing? If no, then it makes sense to add  `CV_DbgAssert()` instead.

@anna-khakimova There are 7 similar loops in this patch. Commit contains 3 updates only. Why?

Reply:
@alalek 
Answer for first question: This checks necessary to process via univ intrinsics last several elements of input array (tail) since their number is less than `nlanes` .
Answer for second question: As I've already mentioned in previous comment, for inplace implementation it's necessary that input and output array types should be the same. So, this check is needed only in 3 cases (in 3 functions which calls when input and output types are the same). 
Note: Please read my comments more attentively.

Reply:
>  Does G-API support inplace processing?

Yes, we _have_ to support inplace processing at least to avoid copy when calling `cv::gapi::wip::draw::render()`.

Not sure if Fluid have ever been tested with such type of inplace execution, but the Fluid itself never forces input and output buffer to be the same.

Reply:
Thanks. Resolved =)

Reply:
Anyway, I would suggest to put `CV_DbgAssert()` for other cases.

> when input and output types are the same

to catch cases when different types are inplaced (e.g. both `float` and `int` are 32-bit)

Reply:
There are no such cases in GAPI.  
For **AbsDiff** there are cases such as:
1. uchar inputs and uchar output
2. short int inputs and short int output.
3. ushort inputs and ushort output.
4. float inputs and float outputs
For all cases mentioned above there is check to detect inplace.

For **Add** and **Sub** there are cases such as:
1. **uchar** inputs and **uchar** output <--- For this case there are check for inplace.
2. **short int** inputs and **uchar** output <---- There is no sense to check for inplace.
3. **float** inputs and **uchar** output <------ There is no sense to check for inplace.
4. **short int** inputs and **short int** output <------- For this case there are check for inplace.
5. **float** inputs and **float** output <-------- For this case there are check for inplace.
6.  **uchar** inputs and **float** output <------- There is no sense to check for inplace.
7. **short int** inputs and **float** output <------- There is no sense to check for inplace.

As you can see, there are no cases with **int** at all and with **int** and **float** in particular.

**These GAPI kernels don't support **int** type at all.**

Inplace implementation is absent for these kernels. **And case when user gives one the same **int** matrix for inputs, then cast this matrix to float and pass it to output is difficult to imagine in reality.**

Reply:
... for now.

Assertions verify assumptions which are required by related code below.
It is a really powerful tool.
This helps with investigations in the future through emitting error messages. This reduces annoying debugging process of related problems.

Reply:
> ... for now.
> 
> Assertions verify assumptions which are required by related code below.
> It is a really powerful tool.
> This helps with investigations in the future through emitting error messages. This reduces annoying debugging process of related problems.

@alalek Ok. Done.

Reply:
> Anyway, I would suggest to put `CV_DbgAssert()` for other cases.
> 
> > when input and output types are the same
> 
> to catch cases when different types are inplaced (e.g. both `float` and `int` are 32-bit)

I've already put checks for inplace when input and output types are the same.

Reply:
> I've already put checks for inplace when input and output types are the same.

Please open an issue internally to start checking for inplace at the Fluid backend level.
Then we could remove such checks from the kernels (where it may be costly given that kernels are called for every line of an image)